### PR TITLE
Initial commit for implementation of Affinity Propagation

### PR DESCRIPTION
**Pull request by Yuan-Ru-Lin**
Implement interface of Affinity Propagation to MLJ. The implementation mimics that of KMedoids and directly uses the existing `MMI.predict`

### CODE REVIEW

**Comment by ablaom** [comment]
@Yuan-Ru-Lin I see this is a draft. Please go ahead and ping me when you are ready for a review. No hurry.

**Comment by Yuan-Ru-Lin** [comment]
The code is ready to be reviewed. I basically just made the model `Static`, turned `MMI.transform` and `MMI.fit` into `MMI.predict` and added doc.

Two dependencies are added: `LinearAlgebra` and `StatsBase` for we need `diagind`, `triu!` and `median` to fill the diagonal part of the similarity matrix with the median similarity of all pairs. I am not familiar with handling `compat` so it may need your attention there.

Oh I should mention that I use `GLMakie` in the doc since it makes visualization easier. I'm not sure where I should add that as a dependency.

**Comment by Yuan-Ru-Lin** [comment]
@ablaom Forgot to ping you!

File: Project.toml [hunk 1]
```diff
@@ -6,10 +6,14 @@ version = "0.1.11"
...
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MLJModelInterface = "e80e1ace-859a-464e-9ed9-23947d8ae3ea"
+StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
 
 [compat]
 Clustering = "0.15"
 Distances = "0.9, 0.10"
+LinearAlgebra = "1.11.0"
 MLJModelInterface = "1.4"
+StatsBase = "0.34.3"
```
Comment (ablaom) [inline Project.toml]:
```suggestion
LinearAlgebra = "1"
MLJModelInterface = "1.4"
StatsBase = "0.34"
```

File: src/MLJClusteringInterface.jl [hunk 2]
```diff
@@ -614,4 +678,73 @@ report(mach).cutter(h = 2.5)
...
+
+- `damp = 0.5`: damping factor
+
+- `maxiter = 200`: maximum number of iteration
+
+- `tol = 1e-6`: tolerance for converenge
+
+- `preference = nothing`: the value of the diagonal elements of the similarity matrix
+
+- `metric = SqEuclidean`: metric (see `Distances.jl` for available metrics)
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]:
```suggestion
- `metric = Distances.SqEuclidean()`: metric (see `Distances.jl` for available metrics)
```

File: src/MLJClusteringInterface.jl [hunk 3]
```diff
@@ -614,4 +678,73 @@ report(mach).cutter(h = 2.5)
...
+
+# Hyper-parameters
+
+- `damp = 0.5`: damping factor
+
+- `maxiter = 200`: maximum number of iteration
+
+- `tol = 1e-6`: tolerance for converenge
+
+- `preference = nothing`: the value of the diagonal elements of the similarity matrix
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]: I guess this is correct, yes?

```suggestion
- `preference = nothing`: the value (single float) of the diagonal elements of the similarity matrix
```

Also, could you please document how an unspecified `preference` is handled. A reference to the wiki article is fine.

**Comment by ablaom** [comment]
@Yuan-Ru-Lin  This is looking great. Thanks for taking the time to wrap your head around the MLJ API and the docstring requirements. 

Can you add some basic test that include:

-  check the report contains what you think it does
- generic interface test (just add `AffinityPropagation` [here](https://github.com/JuliaAI/MLJClusteringInterface.jl/blob/7d0453c583e169c192bad603d1fe15dc1ae2eb24/test/runtests.jl#L154).
- test that the `preference=nothing` fallback has expected behaviour. 

You can bump Julia support in the Project to 1.10 (hopefully fixes your fail) and, if you know how, do the same bump in .github/workflows/ci.yaml - otherwise I can do it.

File: src/MLJClusteringInterface.jl [hunk 4]
```diff
@@ -614,4 +678,73 @@ report(mach).cutter(h = 2.5)
...
+- cluster_labels: labels of clusters given to each datum in `X`
+
+- iterations: the number of iteration run by the algorithm
+
+- converged: whether or not the algorithm converges by the maximum iteration
+
+# Examples
+
+```
+using MLJ, MLJClusteringInterface
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]: The use of `@load` makes loading MLJClusteringInterface redundant:

```suggestion
using MLJ
```

**Comment by Yuan-Ru-Lin** [comment]
Hi @ablaom, I have

- added tests for the model
- bumped supported Julia version from 1.6 to 1.10 in both Project.toml and ci.yml
- edit docs as you suggested.

**Comment by Yuan-Ru-Lin** [comment]
@ablaom Morning from PST! The PR is ready to be reviewed again.

**Review by ablaom — APPROVED** [review]
Thanks again @Yuan-Ru-Lin  for this valuable contribution.

**Pull request merged**